spec/unit/logstash_output_spec.rb

```diff
@@ -184,10 +194,161 @@
+          before(:each) do
+            expect(registered_plugin.http_client).to receive(:post).and_raise(manticore_exception)
+          end
+          it "returns :retry" do
+            expect(transmit_result).to eql :retry
+          end
+        end
+      end
+    end
+  end
+
+  describe '#multi_receive' do
+    let(:events) { [event] }
+
+    context "when first transmit succeeds" do
+      before(:each) do
+        allow(registered_plugin).to receive(:transmit).and_return(:abort).once
+      end
+      it "transmits once" do
+        registered_plugin.multi_receive(events)
+        expect(registered_plugin).to have_received(:transmit).once
+      end
+    end
+
+    context "when first transmit gets terminal failure" do
+      before(:each) do
+        allow(registered_plugin).to receive(:transmit).and_return(:abort).once
+      end
+      it "transmits once" do
+        registered_plugin.multi_receive(events)
+        expect(registered_plugin).to have_received(:transmit).once
+      end
+    end
+
+    context "when transmit indicates that a retry is required" do
+      # Configure a _sequence_ of `#transmit` responses
+      # emits :retry `retry_count` times, invoking `limit_met_hook` (if present) before the
+      # LAST normal retry, then emits `limit_met_next_action`.
+      # as a safeguard, this mock can be called AT MOST `retry_count + 1` times
+      let(:retry_count) { 3 }
+      let(:limit_met_hook) { nil }
+      let(:limit_met_next_action) { :retry }
+      before(:each) do
+        attempts_made = 0
+        next_action = :retry
+        allow(registered_plugin).to receive(:transmit).with(any_args) do
+          current_action = next_action
+          attempts_made += 1
+          if attempts_made == retry_count
+            limit_met_hook&.call
+            next_action = limit_met_next_action
+          end
+          current_action
+        end.at_most(retry_count + 1).times
+      end
+
+      context "and transmit eventually succeeds" do
+        let(:limit_met_next_action) { :done }
+        it "stops retrying" do
+          registered_plugin.multi_receive(events)
+
+          expect(registered_plugin).to have_received(:transmit).exactly(retry_count + 1).times
+        end
+      end
+
+      context "and transmit eventually gets terminal failure" do
+        let(:limit_met_next_action) { :abort }
+        it "stops retrying" do
+          registered_plugin.multi_receive(events)
+
+          expect(registered_plugin).to have_received(:transmit).exactly(retry_count + 1).times
+        end
+      end
+
+      context "and the pipeline shutdown is requested before transmission succeeds" do
+        let(:limit_met_hook) do
+          ->() { expect(registered_plugin).to receive(:pipeline_shutdown_requested?).and_return(:true) }
+        end
+        let(:limit_met_next_action) { :retry }
+
+        if ::Gem::Version.create(LOGSTASH_VERSION) >= ::Gem::Version.create('8.8.0')
```

Comment:
@yaauie faced an issue with `defined?` (Kernel method?) always return `method` which is not our expectation. Tried `Object/Module#const_defined?` which also returns always false with 8.8+ versions. So, similarly with our main logic, using the version condition.

Reply:
🤦🏼 I understand _why_ it doesn't work now. **+1 to your use of version-gating**.

The why:

JRuby's top-level `org` object (and other top-level "normal namespace") is a [`Java::JavaPackage`](https://github.com/jruby/jruby/blob/master/core/src/main/java/org/jruby/javasupport/JavaPackage.java), which responds to _anything_. When it is sent a message that starts out lower-case, it returns another `Java::JavaPackage` with the accumulated namespace, and when it is sent a message that starts out upper-case, it returns a java proxy that _resolves_ the class in the accumulated namespace.

The Ruby `defined?` keyword _effectively_ resolves n-1 of its argument (so when `org.logstash.execution.AbortedBatchException` is provided, it resolves `org.logstash.execution`), and if _that_ resolves successfully it interrogates it to determine if the last thing (in our case `.AbortedBatchException`) will resolve.

But since `org.logstash.execution` is a `Java::JavaPackage` proxy with the accumulated namespace `org`+`logstash`+`execution` that will `Kernel#respond_to?` `AbortedBatchException`, the `defined?` keyword indicates that it is a `method`. It is only when _calling_ this method that `Java::JavaPackage` would fail to resolve the class.

TIL that Ruby's `defined?` doesn't work for java-defined classes. The only _reliable_ mechanism I have been able to find is a rescue clause:

~~~ diff
-if defined?(org.this.that.Another)
+if (org.this.that.Another rescue nil)
~~~